operating_unit_grt_api_sync/models/operating_unit.py

```diff
@@ -0,0 +1,182 @@
+            vals_with_partner.append(data)
+        self.create(vals_with_partner)
+
+    @staticmethod
+    def _filter_branches_data(data, ou_code_company_mapping):
+        return [
+            d
+            for d in data
+            if d["management_id"][:3] in ou_code_company_mapping
+            and d["management_id_level_number"] == 5
+        ]
+
+    @staticmethod
+    def _filter_branches_to_update(recs, data):
+        """Filter the recordset to only include records that need to be updated."""
+        return recs.filtered(
+            lambda mid: mid.valid_from != data[mid.code]["operational_from"]
+            or mid.valid_until != data[mid.code]["operational_until"]
+            or not mid.synced_with_grt
+        )
+
+    def _update_branch(self, data):
+        self.ensure_one()
+        vals = {}
+        if self.valid_from != data["valid_from"]:
+            vals["valid_from"] = data["valid_from"]
+        if self.valid_until != data["valid_until"]:
+            vals["valid_until"] = data["valid_until"]
+        if not self.synced_with_grt:
+            vals["synced_with_grt"] = True
+        if vals:
+            self.write(vals)
+
+    def _process_grt_branch_data(self, data):
+        compatible_companies = self.env["res.company"].search(
+            [("grt_code_prefix", "!=", False)]
+        )
+        ou_code_company_mapping = {
+            c.grt_code_prefix: c.id for c in compatible_companies
+        }
+
+        if data := self._filter_branches_data(  # noqa: E231 E701
+            data, ou_code_company_mapping
+        ):
+            api_data = {
+                d["management_id"]: {
+                    "valid_from": d["operational_from"],
+                    "valid_until": d["operational_until"],
+                    "branch": d["l5_branch"],
+                    "operating_unit": d["l8_operating_unit"],
+                    "country": d["l10_operating_country"],
+                }
+                for d in data
+            }
+            branches = self.search([])
+            branches_data = {
+                b.code: {
+                    "valid_from": datetime.strftime(b.valid_from, "%Y-%m-%d")
+                    if b.valid_from
+                    else None,
+                    "valid_until": datetime.strftime(b.valid_until, "%Y-%m-%d")
+                    if b.valid_until
+                    else None,
+                    "country": b.partner_id.country_id.name,
+                }
+                for b in branches
+            }
+
+            if branches_create_data := {  # noqa: E231 E701
+                k: v for k, v in api_data.items() if k not in branches_data
+            }:
+                self._create_new_branches(branches_create_data, ou_code_company_mapping)
+
+            if branches_update_data := {  # noqa: E231 E701
+                k: v for k, v in branches_data.items() if k in api_data
+            }:
+                api_data = {
```

Comment:
Can you explain why do you create another dict api_data = {) ??
you have a dict created already. I don't see sense to create another one, it increases complexity a lot.